### ISSUE TITLE
TGraph SaveAs .csv

### ISSUE THREAD

**Issue by ferdymercury** [root-project/root#9491]
### Is your feature request related to a problem? Please describe.
When working with collaborators from other disciplines, it is always problematic to share results and raw data created in ROOT in formats other than PNG or PDF.

### Describe the solution you'd like
It would be really helpful if the TGraph and TGraphErrors classes had an option to `->SaveAs(".csv")` so that the result can be immediately imported into Excel, gnuplot, Python or whatever, without the needing to install pyroot, etc.

### Describe alternatives you've considered
```
void DumpGraph(TGraph* gr, const std::string outName)
{
    std::cout << "Info: Creating " << outName << std::endl;
    std::ofstream outfile(outName);
    const Int_t N = gr->GetN();
    double * const x = gr->GetX();
    double * const y = gr->GetY();
    for(Int_t i=0 ; i<N ; ++i)
      outfile << x[i] << "\t" << y[i] << endl;
    outfile.close();
}

void DumpGraphWithErrors(TGraph* gr, const std::string outName)
{
    std::cout << "Info: Creating " << outName << std::endl;
    if(gr->ClassName()!=TString("TGraphErrors")) {
        std::cout << gr->ClassName() << std::endl;
        return;
    }
    std::ofstream outfile(outName);
    const Int_t N = gr->GetN();
    double * const x = gr->GetX();
    double * const y = gr->GetY();
    double * const ex = gr->GetEX();
    double * const ey = gr->GetEY();
    for(Int_t i=0 ; i<N ; ++i)
      outfile << x[i] << "\t" << ex[i] << "\t" << y[i] << "\t" << ey[i] << endl;
    outfile.close();
}
```

### Additional context
https://root-forum.cern.ch/t/how-to-convert-root-graphs-into-dat-format/32184/8
https://root-forum.cern.ch/t/how-to-convert-root-to-txt-file/30534/4

**Comment by couet** [comment]
SaveAs already exist for TGraph and is [inherited from TObject](https://root.cern/doc/master/TObject_8cxx_source.html#l00601). I am not sur we can use it produce .csv files as a such format will not work for any object .

**Comment by ferdymercury** [comment]
Thanks for the feedback. Could the function be overriden ?
```
TGraph::SaveAs(...)
{
   if (std::ends_with(".csv"))
  {
    //do something
  }
  else
  {
     TObject::SaveAs(...);
  }
}
```

Or alternatively, in TObject::SaveAs, do the following:

```
if ends_with('.csv') and ClassName() == "TGraph")
{
... do something
}
else
{
...do what does now
}
```

**Comment by couet** [comment]
Yes the first approach might be the simplest.
In your example you separate value with tabs. I thought `.csv` meant "comma separated value" therefore should we better do:
```
      outfile << x[i] << "," << y[i] << std::endl;

```
?

**Comment by ferdymercury** [comment]
Maybe we could infer the "delimiter" based on the given extension:
.csv --> comma
.tsv --> tab
.txt --> space

**Comment by couet** [comment]
yes any convention can be decided but I do not think `.tsv` is a known standard extension/format ? At the beginning of this issue you said:  _so that the result can be immediately imported into Excel, gnuplot, Python or whatever,_
Importing in Excel means having a real `.csv`. Do the other app you mentioned need different delimiter ?

**Comment by ferdymercury** [comment]
Thanks for the reply.

tsv is more or less standard, see https://en.wikipedia.org/wiki/Tab-separated_values

Excel can import all three variants easily. 

I would use 'spaces' rather than commas if I wanted later to read from a plain C++ program, because the file is much easier to read then.
You can use a oneliner `while(file >> x >> y)`. If you have commas, then you have do `f.ignore()` or things like that.

**Comment by couet** [comment]
Ok so the 3 extensions make sense. I'll go further with that.

**Comment by couet** [comment]
PR here : https://github.com/root-project/root/pull/10365

**Comment by ferdymercury** [comment]
> PR is now merged.

Thanks so much!

> @lmoneta suggested that we could also think of some more general like a new class `THistSaveAs` alloying to save not only graphs but also histograms. The API would be stil via SaveAs that why this PR is a 1st step toward a wider implementation.

Sounds like a great idea!
